samcli/lib/package/utils.py

```diff
@@ -242,28 +263,55 @@ def make_zip(file_name, source_root):
-                            # Originally set to 0005 in the discussion below
-                            # https://github.com/aws/aws-sam-cli/pull/2193#discussion_r513110608
-                            # Changed to 0755 due to a regression in https://github.com/aws/aws-sam-cli/issues/2344
-                            # Mimicking Unix permission bits and recommanded permission bits
-                            # in the Lambda Trouble Shooting Docs
-                            info.external_attr = 0o100755 << 16
+                    with open(full_path, "rb") as data:
+                        file_bytes = data.read()
+                        info = zipfile.ZipInfo(relative_path)
+                        # Context: Nov 2020
+                        # Set external attr with Unix 0755 permission
+                        # Originally set to 0005 in the discussion below
+                        # https://github.com/aws/aws-sam-cli/pull/2193#discussion_r513110608
+                        # Changed to 0755 due to a regression in https://github.com/aws/aws-sam-cli/issues/2344
+                        # Final PR: https://github.com/aws/aws-sam-cli/pull/2356/files
+                        if any(permission_mappers):
                             # Set host OS to Unix
                             info.create_system = 3
+                            # Set current permission of the file/dir to ZipInfo's external_attr
+                            info.external_attr = os.stat(full_path).st_mode << 16
+                            for permission_mapper in permission_mappers:
+                                if permission_mapper:
+                                    info = permission_mapper.apply(info)
+                            # Set current time to be the last time the zip content was modified.
+                            info.date_time = time.localtime()[0:6]
                             zf.writestr(info, file_bytes, compress_type=compression_type)
-                    else:
-                        zf.write(full_path, relative_path)
+                        else:
+                            zf.write(full_path, relative_path)
 
     return zipfile_name
 
 
+make_zip = functools.partial(
+    make_zip_with_permissions,
+    permission_mappers=[
+        WindowsFilePermissionMapper(permissions=0o100755) if platform.system().lower() == "windows" else None,
+        WindowsDirPermissionMapper(permissions=0o100755) if platform.system().lower() == "windows" else None,
+    ],
+)
+# Context: Jan 2023
+# NOTE(sriram-mv): Modify permissions regardless of the Operating system
+# to add 111 for directories and 444 for files in addition to existing permissions.
+# No overriding explicit permissions are set.
+# Extended Attributes are preserved.
+make_zip_with_lambda_permissions = functools.partial(
+    make_zip_with_permissions,
+    permission_mappers=[
+        WindowsFilePermissionMapper(permissions=0o100755) if platform.system().lower() == "windows" else None,
```

Comment:
If we are adding the `WindowsFilePermissionMapper` by checking the OS first, then if statement in the mapper itself seems redundant. Can we get rid of one to make it easy to read?

And also, I would recommend not adding `None` into array (with also supporting it type checking) so that during the execution of these mapper would be simpler, we won't need to have the `if` statement above when we are running the mapper itself.

Reply:
Agree, I was refactoring on the fly, and it can be made better.